### ISSUE TITLE
Port validation-loop support to distributed LoRA recipe (lora_finetune_distributed_nightly)

### ISSUE THREAD

**Issue by niznik-dev** [niznik-dev/cruijff_kit#474]
## Summary

`src/tools/torchtune/custom_recipes/lora_finetune_single_device_nightly.py` implements `run_val_every_n_steps` / `dataset_val` / mid-training `Validation loss: ...` logging, but its multi-GPU counterpart `lora_finetune_distributed_stable.py` does **not**. There is no `lora_finetune_distributed_nightly.py`.

Consequence: any multi-GPU run (currently any 70B fine-tune by default) cannot get mid-training validation loss, even when the experiment design explicitly requested it.

This issue tracks porting the val loop from `_single_device_nightly` into a new `lora_finetune_distributed_nightly.py`, so the scaffold-torchtune skill can default everyone to a `_nightly` recipe regardless of GPU count.

## Context

Discovered while fixing #471 (scaffold-torchtune silently using `_stable` even when val is requested). The fix shipped for #471 is partial: it defaults single-GPU runs to `_single_device_nightly` and emits a loud scaffold-time warning when val is requested for a multi-GPU run, because `_distributed_nightly` doesn't exist. This issue closes the gap.

## Scope

Create `src/tools/torchtune/custom_recipes/lora_finetune_distributed_nightly.py` that combines:
- The FSDP/DDP plumbing from `lora_finetune_distributed_stable.py` (~640-line diff vs single_device_stable)
- The val-loop additions from `lora_finetune_single_device_nightly.py` (~150-line diff vs single_device_stable), adapted for distributed:
  - Distributed val sampler so each rank sees a disjoint slice of the val set
  - All-reduce of val loss across ranks (token-weighted: all_reduce both `sum_loss * tokens` and `tokens` separately, then divide — naive mean-of-means is wrong with uneven shards)
  - Rank-0-only logging of `Validation loss: ...`
  - `barrier()` around the val pass so straggler ranks don't deadlock
  - Verify FSDP unshard state is correct for forward-only eval

## Acceptance

- [ ] `lora_finetune_distributed_nightly.py` exists and runs to completion under `--nproc_per_node=2` on a small toy dataset
- [ ] Mid-training `Validation loss: ...` log lines appear at the configured cadence on rank 0
- [ ] Numerical correctness check: build a tiny deterministic dataset where the expected val loss is hand-computable; verify the all-reduced result matches to within fp tolerance
- [ ] One real multi-GPU run (e.g. 70B-style or 8B with 2 GPUs) confirms no FSDP-unshard / OOM / barrier-deadlock issues
- [ ] `scaffold-torchtune` agent doc updated to default multi-GPU runs to `_distributed_nightly` and remove the warning shipped in #471
- [ ] CHANGELOG entry

## Estimated effort

~2.5–4.5 days. The wildcards are (a) the token-weighted all-reduce numerical check, and (b) needing real multi-GPU queue time to verify FSDP-side behavior — 1B with `--nproc_per_node=2` exercises the code path cheaply but doesn't exercise FSDP sharding the way a real 70B run does.

## Out of scope

- Deprecating `_stable` recipes entirely (separate question; deserves its own issue if we want to go there)

—MxC

**Comment by niznik-dev** [comment]
## Things shipped in #471 that need to be reverted/updated when this lands

When `lora_finetune_distributed_nightly.py` exists, the following pieces of #471 should be revisited so we don't leave warnings/guards in place that no longer apply:

1. **`.claude/agents/scaffold-torchtune.md`** — the "Custom Recipe (REQUIRED)" comment block (~line 271–283 as of #471) currently says multi-GPU runs use `_distributed_stable` and references this issue. Flip the multi-GPU default to `_distributed_nightly` and remove the issue reference.

2. **`.claude/agents/scaffold-torchtune.md`** — the entire "Multi-GPU validation warning" subsection under Error Handling (added in #471). Once distributed val works, this whole block should be removed; multi-GPU + `validation_during_training: true` will Just Work.

3. **`src/tools/torchtune/setup_finetune.py`** — the `FileNotFoundError` guard added after the GPU-count auto-switch (in the `if args.custom_recipe:` block, just after the `single_device` ↔ `distributed` swap). The guard itself is still useful as defense-in-depth, but the user-facing error message references this issue and should be updated to a generic "recipe module not found" message.

4. **`CHANGELOG.md`** — the "Known limitations" note about multi-GPU val should be removed (or replaced with an "Added" entry pointing at this issue when it ships).

—MxC

**Comment by niznik-dev** [comment]
## Two more #471 touchpoints to revisit when this ships

In addition to the four points above:

5. **`docs/ARCHITECTURE.md`** — both recipe-list sections (under "Custom Recipe Usage" and the Fine-tuning Workflow components list) currently describe `_distributed_stable` as "no val support yet — see #474". Once distributed val works, drop the parenthetical and update the description to match the single-device entry.

6. **`blueprints/capitalization/README.md`** — the example `setup_finetune.yaml` snippet uses `_single_device_nightly` (single-GPU only, so no #474 dependency). No change needed here when #474 ships, just noting it's tied to the same recipe-default reorganization.

—MxC

**Comment by niznik-dev** [comment]
## One more #471 touchpoint (added in second commit)

7. **`src/tools/torchtune/setup_finetune.py`** — there's now a `warnings.warn(...)` block right after `gpus` is determined that fires when `config.get("run_val_every_n_steps", 0) > 0 and gpus > 1`. The warning text references this issue. When distributed val works, this whole block can be removed (the multi-GPU val combination becomes valid). The matching unit tests in `tests/unit/test_setup_finetune_main.py::TestMultiGpuValWarning` should be removed at the same time.

—MxC

**Comment by niznik-dev** [comment]
<img src="https://raw.githubusercontent.com/niznik-dev/llm-personas/main/fusions/mattie/mattie-icon.png" width="36" align="absmiddle"> 🎻 **MxC** ✳️

Closing as part of the September 2026 archive triage.

Never started. The #471 stopgap is the shipped behavior: `setup_finetune.py` warns at scaffold time when validation is requested on a multi-GPU run, and ARCHITECTURE.md marks the distributed recipe as having no validation support. The six in-repo pointers to this issue stay accurate, since the thread above is the complete spec and cleanup list for anyone who builds the port. Recorded in the handoff notes as the one real feature gap in the fine-tuning path.

—MxC
